manticore/ethereum.py

```diff
@@ -1131,44 +1142,76 @@ def global_findings(self):
+        space_for_all_size_fields = number_dyn_arguments * 32
 
+        # TODO FIXME (mark) what if space_for_arg_data or space_for_each_arg ends up being zero?
+        space_for_arg_data = len(data) - free_pointer - space_for_all_size_fields
 
-        #This will try certain partition of the data into arguments.
-        #It may generate an unsolvable core. Other feasible partitions may exist. 
-        argument_size = available/number_dyn_arguments
-        for index in dyn_arguments:
-            #get, constraint and concretize dyn_offset to some reasonable value
-            dyn_offset = ABI.get_uint(data, 4 + index*32, 256)
-            state.constrain(dyn_offset+4 == free_pointer)
-            data[4 + index*32 : 4 + index*32 + 32] = ("%064x"%(free_pointer-4)).decode('hex')
 
+        #
+        # This will try certain partition of the data into arguments.
+        # It may generate an unsolvable core. Other feasible partitions may exist.
+        #
 
-            #get, constraint and concretize dyn_size to some reasonable value
-            dyn_size = ABI.get_uint(data, free_pointer, 256)
-            state.constrain(dyn_size == argument_size/32)
-            data[free_pointer:free_pointer+32]= ("%064x"%(argument_size/32)).decode('hex')
+        # space_for_each_arg needs to be a multiple of 32, as required by the ABI. so if the math didn't work
+        # out cleanly, we force it, and round it down to the next multiple of 32. this does waste some space,
+        # meaning not every byte in the data will correspond to an argument.
+
+        space_for_each_arg = (space_for_arg_data/number_dyn_arguments) & (~0x1f)
+        assert space_for_each_arg % 32 == 0
+
+        for index in dyn_arguments:
+            # Constrain/concretize the argument's head element, which can be calculated based on the number
+            # of dynamic arguments in the type spec, as the length of the data
+            arg_head_element_offset = 4 + index * 32
+            offset_to_arg_data = ABI.get_uint(data, 32, arg_head_element_offset)
+
+            concrete_offset_to_arg_data = free_pointer - 4
+            state.constrain(offset_to_arg_data == concrete_offset_to_arg_data)
+            data[arg_head_element_offset:arg_head_element_offset + 32] = ("%064x"%(concrete_offset_to_arg_data)).decode('hex')
+
+            # Constrain/concretize the argument's size field, which we calculate based on the number of
+            # dynamic arguments and the total amount of space they all share
+            number_of_elements_in_arg = ABI.get_uint(data, 32, free_pointer)
+
+            # FIXME TODO (mark) i'm not sure that the encoding for bytes and string is identical, this could
+            # be buggy for string
+            typ = types[index]
+            if typ not in ('bytes', 'string'):
+                element_size = 32
+            else:
+                element_size = 1
 
+            concrete_number_of_elements_in_arg = space_for_each_arg/element_size
```

Comment:
In a future? We should ask the solver for a plausible solution instead of forcing/attempting a specific size which may not be feasible.